src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageSource.java

```diff
@@ -0,0 +1,539 @@
+		if (this.consumer != null) {
+			Consumer<K, V> consumer2 = this.consumer;
+			this.consumer = null;
+			synchronized (this.consumerMonitor) {
+				consumer2.close(30, TimeUnit.SECONDS);
+			}
+		}
+	}
+
+	/**
+	 * AcknowledgmentCallbackFactory for KafkaAckInfo.
+	 *
+	 * @param <K> the key type.
+	 * @param <V> the value type.
+	 *
+	 */
+	public static class KafkaAckCallbackFactory<K, V> implements AcknowledgmentCallbackFactory<KafkaAckInfo<K, V>> {
+
+		@Override
+		public AcknowledgmentCallback createCallback(KafkaAckInfo<K, V> info) {
+			return new KafkaAckCallback<K, V>(info);
+		}
+
+	}
+
+	/**
+	 * AcknowledgmentCallback for Kafka.
+	 *
+	 * @param <K> the key type.
+	 * @param <V> the value type.
+	 *
+	 */
+	public static class KafkaAckCallback<K, V> implements AcknowledgmentCallback, Acknowledgment {
+
+		private final Log logger = LogFactory.getLog(getClass());
+
+		private final KafkaAckInfo<K, V> ackInfo;
+
+		private volatile boolean acknowledged;
+
+		private boolean autoAckEnabled = true;
+
+		public KafkaAckCallback(KafkaAckInfo<K, V> ackInfo) {
+			this.ackInfo = ackInfo;
+		}
+
+		@Override
+		public void acknowledge(Status status) {
+			Assert.notNull(status, "'status' cannot be null");
+			if (this.acknowledged) {
+				throw new IllegalStateException("Already acknowledged");
+			}
+			synchronized (this.ackInfo.getConsumerMonitor()) {
+				try {
+					ConsumerRecord<K, V> record = this.ackInfo.getRecord();
+					switch (status) {
+					case ACCEPT:
+					case REJECT:
+						commitIfPossible(record);
+						break;
+					case REQUEUE:
+						rollback(record);
+						break;
+					default:
+						break;
+					}
+				}
+				catch (WakeupException e) {
+					throw new IllegalStateException(e);
+				}
+				finally {
+					this.acknowledged = true;
+					if (!this.ackInfo.isAckDeferred()) {
+						this.ackInfo.getOffsets().get(this.ackInfo.getTopicPartition()).remove(this.ackInfo);
+					}
+				}
+			}
+		}
+
+		private void rollback(ConsumerRecord<K, V> record) {
+			this.ackInfo.getConsumer().seek(this.ackInfo.getTopicPartition(), record.offset());
+			Set<KafkaAckInfo<K, V>> inflight = this.ackInfo.getOffsets().get(this.ackInfo.getTopicPartition());
+			if (inflight.size() > 1) {
+				List<Long> rewound =
+						inflight.stream()
+							.filter(i -> i.getRecord().offset() > record.offset())
+							.peek(i -> {
+								i.setRolledBack(true);
+							})
+							.map(i -> i.getRecord().offset())
+							.collect(Collectors.toList());
+				if (rewound.size() > 0 && this.logger.isWarnEnabled()) {
+					this.logger.warn("Rolled back " + record + " later in-flight offsets "
+							+ rewound + " will also be re-fetched");
+				}
+			}
+		}
+
+		private void commitIfPossible(ConsumerRecord<K, V> record) {
+			if (this.ackInfo.isRolledBack()) {
+				if (this.logger.isWarnEnabled()) {
+					this.logger.warn("Cannot send offset for " + record
+							+ " to transaction; an earlier offset was rolled back");
```

Comment:
This message is somehow not related any more. We don't provide any transaction functionality explicitly.